crates/wasmtime/src/externals.rs

```diff
@@ -864,6 +902,38 @@ impl Memory {
         MemoryType::from_wasmtime_memory(&self.wasmtime_export.memory.memory)
     }
 
+    /// Copies memory contents at the given offset into a buffer.
+    ///
+    /// Returns an error if the copy failed.
+    pub fn read(&self, offset: usize, buffer: &mut [u8]) -> Result<(), MemoryError> {
+        if offset + buffer.len() > self.data_size() {
```

Comment:
I think here this will want to use `checked_add` to avoid overflow.

Reply:
Ah yeah, I'm too used to `debug` semantics for overflow checking...

Both `read` and `write` now use `checked_add`.